### ISSUE TITLE
mongodb version 2.2.31 leaks memory

### ISSUE THREAD

**Issue by andrasq** [mongodb-js/mongodb-core#220]
with the 2.2.31 version of the mongodb driver the node rss and heap usage keeps increasing; with 2.1.18 heap usage was level.

Investigating, we found hundreds of thousands of strings being retained by `_hiCache` looking like eg `"ee046b1bf889659f992e1bf5792697b6_+b22YJWKCCSauNC3hATvVA==_5000"`, and hundreds of thousands of 80-byte `Uint8Arrays` keyed by these strings.  They look like the keys and data held by `_hiCache` in `lib/auth/scram.js`.  Searching the sources, it doesn't look like this cache is ever emptied.

**Comment by andrasq** [comment]
Possibly related, we are also seeing 4x more cpu used per call on the server running 2.2.31 than 2.1.18, so it's possible caching didn't change but the driver authenticates more frequently than it used to.

**Comment by mbroadst** [comment]
@andrasq Hi! Sounds like a pretty severe bug, and most likely related to 07d68bc87485b8bab710c9e1387b24d098b3169c. Would it be possible for you to measure the performance by simply disabling the caching mechanism introduced? (comment out lines 80-83 & 87)

**Comment by andrasq** [comment]
thanks for the PR link!  I timed the old and new implementations standalone, and the new is over 5x faster (on my machine, 4500 MHz i7, the old `hmac digest` is 70 hashes/sec, new `crypto.pbkdf2Sync` is 375 / sec).  So even without caching it's much faster than what it replaced.

Do you happen to know if auth is getting called more often than before?  Still puzzled about the increased cpu, especially since the new code is so much faster.

**Comment by andrasq** [comment]
over an hour of runtime the cpu usage without `_hiCache` is about the same (within 1% of with `_hiCache`)

**Comment by mbroadst** [comment]
@andrasq the answer is almost certainly "yes" as you can see from the [diff here](https://github.com/mongodb-js/mongodb-core/compare/V1.3.8...v2.1.15) there were _many_ changes in that time.  Unfortunately it was before my time, so I don't know off the top of my head.  If you file a ticket on our JIRA we can try to investigate and incorporate into the upcoming 3.x release - or certainly any contributions are welcome!

**Comment by andrasq** [comment]
on the same production machine that exhibited the memory leak, the modified 2.2.31 driver with `_hiCache` commented is now indistinguishabe from the 2.1.18 version -- there is no visible latency or cpu impact of not reusing the ScramSHA1 hashes.  No mysterious increased cpu usage, either.

What was the use case that motivated the inclusion of `_hiCache`?  Could this be as simple as just removing it?

**Comment by mbroadst** [comment]
@andrasq yes we should just remove it, I think it was premature optimization on the part of the submitter, and a spotty review on my part :see_no_evil:

**Comment by andrasq** [comment]
@mbroadst see above re performance by simply disabling the caching

**Comment by andrasq** [comment]
awesome, thanks!  Do you have an ETA?  Do you need a PR?

**Comment by mbroadst** [comment]
@andrasq a PR would be most welcome! I can publish a release in a few days in that case, as I was waiting for resolution on this ticket before publishing a few other changes.

**Comment by andrasq** [comment]
PR #223

**Comment by owenallenaz** [comment]
I added the `_hiCache` in a pull request earlier. According to my tests your mongo server should re-use the same _hiCache entry for each server you connect to. We've had this running in our environment for a long time and have noticed no memory issues at all. When I connect to our environment we generate one entry in the _hiCache per user, per database that we connect on. So in our case our clients boot with 8 pool members on one db, and 2 pool members on another (across 3 repl set members). I get a total of 2 _hiCache entries. We've tried rebooting our mongo servers, doing primary switches, none of those cause new entries in _hiCache. The cost of generating that cache entry is not trivial. In the pull request I added that code block alone was taking seconds to complete, and it went down to milliseconds. We've seen a significant boost in boot performance since that's gone into affect.

I would urge caution simply pulling that out without further looking into the exact circumstances that appear to be causing that. My guess is that there is something else happening in this case which is causing the salt to not be re-used or something.

**Comment by andrasq** [comment]
Don't know why it's not reusing the _hiCache entries, we run a fairly generic setup with 1 primary and 2 replicas.  There were no reboots or primary changes during our test run, just steady heap and rss growth.

Pulling out the hash function eliminates the memory leak at a cost of 3.6 ms per hash (`crypto. pbkdf2Sync` call on an AWS E5-2676 v3 @ 2.40GHz).  The old non-caching hand-rolled loop took 22 ms, so even without caching the new code is saving 18.4 ms per call.  (A 100 ms startup savings with only 2 _hiCache entries suggests that the old hash function might have been used.)

**Comment by mbroadst** [comment]
@owenallenaz I think its still a valid criticism that the cache grows unbounded.  I'd have to dig further into this to see what's going on between your two deployments, but perhaps a compromise for the short term is to make the cache a bit smarter ensuring that items are evicted in some reasonable time period (perhaps an LRU).  There is enough traffic through this object that we could provide reasonable book keeping.

Regardless, I'm hesitant to simply reintroduce the concept without more benchmarking - reverting to the current form reaches the original goal of the first PR (replace the custom hasher with a far more performant version). To reiterate: I'm not against including the cache, perhaps you and @andrasq could work together to some better general solution?

**Comment by andrasq** [comment]
@mbroadst it would be simpler to just purge the cache after every 200 entries.  This would capture a good working set, would cap the amount of memory used, and would be faster and simpler than an LRU cache.

```
var _hiCache = {},
    _hiCacheCount = 0;
function _hiCachePurge() { _hiCache = {}; _hiCacheCount = 0 }

// insert new key
if (_cacheCount >= 200) _hiCachePurge();
_hiCache[key] = data;
_hiCacheCount += 1;
```

**Comment by mbroadst** [comment]
@andrasq that sounds reasonable as well.  I'm not sure if these micro optimizations help that much anymore but you used to be able to get better performance out of Object's as a cache by using e.g. `_hiCache = Object.create(null)` and then using a stored copy of `Object.prototype.hasOwnProperty` to look up keys on it.  I have very limited time at the moment, but am comfortable considering a solution if you are willing to put it together and test in your environment.

**Comment by bompus** [comment]
The suggestion by @andrasq seems reasonable to me. I'd still be curious of the environment setup where you were able to see many items in the cache, but this seems like a decent middle ground to still keep the cache, but limit it.

@mbroadst - This can probably also close out the pending JIRA issue to implement cache in this driver ( https://jira.mongodb.org/browse/NODE-990 )

**Comment by mbroadst** [comment]
@bompus indeed :) Glad you saw the ticket, we have two new people joining the team here at Mongo and this was going to be one of their first projects. 

@andrasq could you provide more details about your deployment, so that we can try to simulate locally?  Ideally we'd be able to write some unit tests to express the issue.

**Comment by andrasq** [comment]
see PR #225 for an implementation of the capped-at-200 suggestion

I also spent some time benchmarking various has-property tests, and ended up not changing the `!== undefined`: losing the clarity of the comparison for the 25%-35% speed difference is not worth it here (20 mill tests / sec vs 25 m/s; larger gains on older node eg v5.8.0 and v4.4.0)

(Interesting aside:  the fastest was using `.hasOwnProperty` on a `new Map`, then on `{}` and `new Object()`.  `Object.create(null)` is a bit slower still (except in a few cases like node v5.8.0 and v9.x).  Calling a stored `hasOwnProperty` function consistently trailed `Object.prototype.hasOwnProperty.call` by a hair (2-3%).  I did not time setting and getting properties, though in actual usage it's the overall time that matters.)

**Comment by andrasq** [comment]
@mbroadst re deployment configs, I noticed that the patched (cache-less) 2.1.15 mongodb-core lowers cpu usage compared to 1.3.18 during our hourly system script runs which make a large number of `collStats` calls with `{ readPreference: 'secondaryPreferred' }`.  It may be that this is where the extra authentication is occurring, since the new salted password hash is known to use a lot less cpu.

I also coincidentally discovered that `admin.system.users` has a lot of leftover rows with credentials to db's that no longer exist, and that these old db names occur as substrings of the names being collStats-ed.  Not sure if this is relevant though.

**Comment by mbroadst** [comment]
@andrasq @bompus thanks for your work on this.  The change has been merged in with the cap of 200 entries, and will go out in a release this week after I write some tests for #224
